### ISSUE TITLE
response_format can return successful output that violates the requested JSON shape

### ISSUE THREAD

**Issue by Thump604** [waybarrios/vllm-mlx#635]
# `response_format` can return successful output that violates the requested JSON shape

## Reproduction

Run `vllm-mlx` with a Qwen 3.6 model and send streaming chat-completion requests
with either:

1. `response_format={"type":"json_object"}` and a prompt requesting
   `[ {"ok": true} ]`.
2. A strict JSON schema whose root is an object and whose `sc` property uses
   nine homogeneous `prefixItems`, `minItems: 9`, and `maxItems: 9`, with each
   item constrained to an integer from 1 through 5.

The local reproduction used `qwen3.6-35b-a3b`, the upstream v0.4.0 code paths,
and the request bodies preserved in:

- `run/qwen-constrained-output-investigation/20260713T192436Z/pre-fix-json-object-request.json`
- `run/qwen-constrained-output-investigation/20260713T192436Z/pre-fix-json-schema-request.json`

Send either body to `POST /v1/chat/completions`:

```json
{"model":"qwen3.6-35b-a3b","messages":[{"role":"user","content":"Return exactly this JSON array and nothing else: [{\"ok\":true}]"}],"stream":true,"max_tokens":256,"temperature":1.0,"top_p":0.95,"top_k":20,"min_p":0.0,"presence_penalty":0.0,"repetition_penalty":1.0,"response_format":{"type":"json_object"}}
```

```json
{"model":"qwen3.6-35b-a3b","messages":[{"role":"user","content":"Return a compact QA object. Use p=false, s=review, nine scores of 4, and no issues. Return JSON only."}],"stream":true,"max_tokens":512,"temperature":1.0,"top_p":0.95,"top_k":20,"min_p":0.0,"presence_penalty":0.0,"repetition_penalty":1.0,"response_format":{"type":"json_schema","json_schema":{"name":"qa","strict":true,"schema":{"type":"object","additionalProperties":false,"required":["p","s","sc","i"],"properties":{"p":{"type":"boolean"},"s":{"type":"string","maxLength":240},"sc":{"type":"array","minItems":9,"maxItems":9,"prefixItems":[{"type":"integer","minimum":1,"maximum":5},{"type":"integer","minimum":1,"maximum":5},{"type":"integer","minimum":1,"maximum":5},{"type":"integer","minimum":1,"maximum":5},{"type":"integer","minimum":1,"maximum":5},{"type":"integer","minimum":1,"maximum":5},{"type":"integer","minimum":1,"maximum":5},{"type":"integer","minimum":1,"maximum":5},{"type":"integer","minimum":1,"maximum":5}]},"i":{"type":"array","maxItems":3,"items":{"type":"object","additionalProperties":false,"required":["a","v","c","m","e","r"],"properties":{"a":{"type":"string","enum":["r","c","p"]},"v":{"type":"string","enum":["b","w","i"]},"c":{"type":"string","maxLength":48},"m":{"type":"string","maxLength":280},"e":{"type":"array","maxItems":4,"items":{"type":"string","maxLength":80}},"r":{"type":"array","maxItems":4,"items":{"type":"string","maxLength":80}}}}}}}}}}
```

## Observed behavior

- `json_object` accepted an array root and completed with
  `finish_reason=stop`.
- The strict schema completed with `finish_reason=stop` while `sc` contained
  strings instead of nine bounded integers.
- A processor setup or runtime failure could disable constraints and continue
  generation.
- Streaming validation happened after response headers/chunks had already been
  exposed, so invalid output could no longer fail the HTTP request cleanly.

## Expected behavior

- `json_object` permits only an object root.
- `json_schema` either enforces the requested semantics or rejects unsupported
  schema shapes before generation.
- Processor failures do not silently downgrade to unconstrained generation.
- A structured stream is independently validated before HTTP success is
  exposed.

## Compared upstream code paths

- `vllm_mlx/constrained/json_schema_processor.py`
- `vllm_mlx/api/tool_calling.py`
- `vllm_mlx/server.py`

The current v0.4.0 processor uses an object-or-array generic grammar,
simplifies away unsupported `prefixItems` semantics, treats syntactically
complete JSON as sufficient for EOS, and contains unconstrained fallback paths.
The streaming server validates only after yielding output.

## Minimal patch/design ask

1. Require an object root for `json_object` in grammar and post-validation.
2. Preserve homogeneous bounded `prefixItems`; reject tuple/open-ended forms
   when their semantics cannot be preserved.
3. Gate EOS on independent schema validity.
4. Fail closed on constrained-decoder setup/runtime errors.
5. Buffer structured streams until final independent validation succeeds.
6. Add regression tests for the exact compact schema and object-root case.

This follows the hard-fail direction established by #539 and is adjacent to,
but distinct from, the stall report in #546.

## Explicit non-claims

This does not claim a Qwen model-quality failure, general Qwen instability, or
that every JSON Schema keyword can be represented by `lm-format-enforcer`.
Unsupported semantics should be rejected rather than advertised as strict.

**Comment by Thump604** [comment]
Follow-up testing exposed a second correctness defect in the same constrained-output path.

## Exact reproduction

The failing request used streaming `response_format=json_schema` with an object containing `sc`, an array of exactly nine integers, each constrained to `minimum: 1` and `maximum: 5`. The prompt and schema were preserved locally in:

`/opt/ai-runtime/run/qwen-constrained-output-investigation/20260713T-numeric-range/work-item-1170-request.json`

The model selected nine `0` values. The existing processor allowed those tokens, then reached an empty allowed-token set after the invalid array had already been generated and returned HTTP 500.

## Compared code path

The affected construction path is `vllm_mlx/api/tool_calling.py::build_json_logits_processor` into `vllm_mlx/constrained/json_schema_processor.py`. The existing `lm-format-enforcer` path did not enforce the declared numeric bounds or fixed array length at token-selection time.

## Patch

PR #636 now routes strict `json_schema` requests through a request-local `llguidance` matcher while retaining the existing LMFE path for `json_object`. It also handles padded MLX vocabularies, schema-aware EOS, and bounds non-progress whitespace outside JSON strings.

Focused regressions prove that a logit preference for `0` is masked when only `1` through `5` are legal, early array closure is masked, padded model-vocabulary slots remain masked, and EOS is accepted only after a complete schema-valid value.

## Validation

- full upstream suite: 2,261 passed, 24 skipped, 23 deselected
- focused structured/server suite: 222 passed, 2 skipped, 3 deselected
- Ruff and formatting checks: passed
- exact 51,014-character live request: 3/3 HTTP 200, `finish_reason=stop`, independently schema-valid, with all nine scores in range
- explicit conflict run requesting score `0`: HTTP 200, `finish_reason=stop`, emitted legal score `1`

Live summary:

`/opt/ai-runtime/run/qwen-constrained-output-investigation/20260713T-live-numeric-range-llguidance-certification-pass/summary.json`

## Explicit non-claims

This does not classify Qwen model quality and does not claim universal JSON Schema support. It fixes token-level enforcement for strict schemas representable by the selected grammar backend; unsupported or invalid setup remains fail-closed.

**Comment by Thump604** [comment]
Resolved by #636, which is merged. Closing as completed.
